mesonbuild/modules/simd.py

```diff
@@ -13,53 +13,77 @@
 # limitations under the License.
 from __future__ import annotations
 
-from .. import mesonlib, compilers, mlog
+import typing as T
+
+from .. import mesonlib, mlog
 from .. import build
+from ..compilers import Compiler
+from ..interpreter.type_checking import SOURCES_KW, STATIC_LIB_KWS
+from ..interpreterbase.decorators import KwargInfo, typed_pos_args, typed_kwargs
 
 from . import ExtensionModule, ModuleInfo
 
+if T.TYPE_CHECKING:
+    from ..interpreter import kwargs as kwtypes
+
+    class CheckKw(kwtypes.StaticLibrary):
+
+        compiler: Compiler
+        mmx: T.List[mesonlib.FileOrString]
+        sse: T.List[mesonlib.FileOrString]
+        sse2: T.List[mesonlib.FileOrString]
+        sse3: T.List[mesonlib.FileOrString]
+        ssse3: T.List[mesonlib.FileOrString]
+        sse41: T.List[mesonlib.FileOrString]
+        sse42: T.List[mesonlib.FileOrString]
+        avx: T.List[mesonlib.FileOrString]
+        avx2: T.List[mesonlib.FileOrString]
+        neon: T.List[mesonlib.FileOrString]
+
+
+# FIXME add Altivec and AVX512.
+ISETS = (
+    'mmx',
+    'sse',
+    'sse2',
+    'sse3',
+    'ssse3',
+    'sse41',
+    'sse42',
+    'avx',
+    'avx2',
+    'neon',
+)
+
+
 class SimdModule(ExtensionModule):
 
     INFO = ModuleInfo('SIMD', '0.42.0', unstable=True)
 
     def __init__(self, interpreter):
         super().__init__(interpreter)
-        # FIXME add Altivec and AVX512.
-        self.isets = ('mmx',
-                      'sse',
-                      'sse2',
-                      'sse3',
-                      'ssse3',
-                      'sse41',
-                      'sse42',
-                      'avx',
-                      'avx2',
-                      'neon',
-                      )
         self.methods.update({
             'check': self.check,
         })
 
-    def check(self, state, args, kwargs):
-        result = []
-        if len(args) != 1:
-            raise mesonlib.MesonException('Check requires one argument, a name prefix for checks.')
-        prefix = args[0]
-        if not isinstance(prefix, str):
-            raise mesonlib.MesonException('Argument must be a string.')
-        if 'compiler' not in kwargs:
-            raise mesonlib.MesonException('Must specify compiler keyword')
+    @typed_pos_args('simd.check', str)
+    @typed_kwargs('simd.check',
+                  KwargInfo('compiler', Compiler, required=True),
+                  *[SOURCES_KW.evolve(name=iset) for iset in ISETS],
```

Comment:
Re-raising this for visibility so we don't forget:

SOURCES_KW is intended and written for declare_dependency, and has the wrong types for library() and friends -- specifically it excludes a couple of things you can add to a library but not to a declared dependency. So I think it's wrong for simd.check as well, we need a new dedicated KwargInfo that aligns with what library() expects and does.

Reply:
Are those all the things that are in SourcesVarargsType?

Reply:
All this FeatureNew discovery is going to be a `****`

Reply:
https://github.com/mesonbuild/meson/pull/12303

Reply:
@tristan957 sources keyword landed. Hopefully with the rebase this will make your life easier